### PR TITLE
std.net: add explicit error sets for IP parsing

### LINKED ISSUE #18276: type weirdness when switching on errors in comptime
### Zig Version

0.11.0, 0.12.0-dev.1821+a38af5f54

### Steps to Reproduce and Observed Behavior

in any comptime context, like global scope
```
const ip = std.net.Address.parseIp("::1", 8090);
```
this produces the following output in 0.11.0:
```
zig build-exe http_test Debug native: error: the following command failed with 1 compilation errors:
/home/user/projects/tar/zig-linux/zig build-exe /home/user/projects/home/http_test/src/main.zig --cache-dir /home/user/projects/home/http_test/zig-cache --global-cache-dir /home/user/.cache/zig --name http_test --listen=- 
Build Summary: 0/3 steps succeeded; 1 failed (disable with --summary none)
install transitive failure
└─ install http_test transitive failure
   └─ zig build-exe http_test Debug native 1 errors
/home/user/projects/tar/zig-linux/lib/std/net.zig:28:13: error: expected type 'error{InvalidCharacter}', found 'error{Overflow}'
            error.Overflow,
            ^~~~~~~~~~~~~~
/home/user/projects/tar/zig-linux/lib/std/net.zig:28:13: note: 'error.Overflow' not a member of destination error set
src/main.zig:3:35: note: called from here
const ip = std.net.Address.parseIp("::1", 8090);
           ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
referenced by:
    server: src/main.zig:24:14
    main: src/main.zig:11:5
    remaining reference traces hidden; use '-freference-trace' to see all reference traces
```

### Expected Behavior

I'd expect this to work in comptime and to return something like described in `@compileLog` as `@as(@typeInfo(@typeInfo(@TypeOf(net.Address.parseIp)).Fn.return_type.?).ErrorUnion.error_set!net.Address, [runtime value])
`

### PR DESCRIPTION
**Pull request by ArtemKolichenkov**
Inferred errors in switch statements prevented IP address parsing at comptime.
Adding explicit error sets fixes it.

Also added a test that checks that:
a) You can parse valid IP address at comptime
b) You can handle error on your end for invalid IP addresses (as opposed to seeing compilation error regarding switch statement)

Closes #18276

### CODE REVIEW

File: lib/std/net/test.zig [hunk 1]
```diff
@@ -4,6 +4,27 @@ const net = std.net;
 const mem = std.mem;
 const testing = std.testing;
 
+test "parse and render IP addresses at comptime" {
```
Comment (ArtemKolichenkov) [inline lib/std/net/test.zig:7]: hm, I wonder if this test should be skipped in WASI? i.e. like in the `parse and render IPv4/6 addresses` tests below?
```zig
if (builtin.os.tag == .wasi) return error.SkipZigTest;
```
Not sure how to test it in WASI. Let me know if it's better to skip for now.

File: lib/std/net.zig [hunk 2]
```diff
@@ -25,20 +38,20 @@ pub const Address = extern union {
...
-            error.Overflow,
-            error.InvalidEnd,
-            error.InvalidCharacter,
-            error.Incomplete,
-            error.NonCanonical,
+            IPv4ParseError.Overflow,
+            IPv4ParseError.InvalidEnd,
+            IPv4ParseError.InvalidCharacter,
+            IPv4ParseError.Incomplete,
+            IPv4ParseError.NonCanonical,
```
Comment (Vexu) [inline lib/std/net.zig:45]: Please don't change these.

Comment (ArtemKolichenkov) [reply lib/std/net.zig:45]: Sure, will roll those back. 👍 
I was questioning whether to change it or not, AFAIK it doesn't affect runtime, so, a style choice. I was looking at other std libs to see if there is a clear convention regarding this, but seems both `error.SomeError` and `SomeErrorSet.SomeError` are used throughout the codebase.
Or is there some edge-cases where two are distinct? Some compiler logs I guess?

File: lib/std/net/test.zig (hunk 1 above)
Comment (ArtemKolichenkov) [reply lib/std/net/test.zig:7]: Ended up adding wasi skip to be in line with other similar tests below.

File: lib/std/net.zig (hunk 2 above)
Comment (Vexu) [reply lib/std/net.zig]: `SomeErrorSet.SomeError` has the type `SomeErrorSet`, `error.SomeError` has the type `error{SomeError}` which implicitly casts to `SomeErrorSet`. `error.SomeError` is preferred because it makes it obvious that the value is an error.

**Comment by ArtemKolichenkov** [comment]
@Vexu CI partially failed (seems like same issue that https://github.com/ziglang/zig/pull/18278 fixes)
Rebased on latest master & force-pushed.
